# Pre-Lecture HW

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
village_df = pd.read_csv(url)
village_df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [6]:
village_df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [9]:

# Check the number of rows and columns
rows, columns = village_df.shape

print(f"The DataFrame has {rows} rows and {columns} columns.")


The DataFrame has 391 rows and 11 columns.


Definitions: <br>
1. Observations: The individual rows in the DataFrame, each representing a single entry in the data set. <br>
2. Variables: The columns in the DataFrame representing a specific characteristic of observations.

In [17]:
# Provides an overview of the DataFrame
village_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


In [18]:
# Provides summary statistics for numerical columns
village_df.describe(include='all')

row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391               391   
unique            92      388               391   
top     K.K. Country  wee one  villager-admiral   
freq              10        2                 1   
mean             NaN      NaN               NaN   
std              NaN      NaN               NaN   
min              NaN      NaN               NaN   
25%              NaN      NaN               NaN   
50%              NaN      NaN               NaN   
75%              NaN      NaN               NaN   
max              NaN      NaN               NaN   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [19]:
df.shape


(391, 11)

The first coordinate in df.shape determines the number of rows (391) and the second coordinate determines the number of columns (11).

In [20]:
df.describe ()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

df.describe () only reported results for one column because that is the only common with numerical values instead of words.

The count column has 391 values, and it doesn't have any missing values.

The difference between an attribute (df.shape) and a method (df.describe ()) is that the attribute df.shape provides a property of the object directly without requiring parentheses because it's not a function and can simply return the property of the object. The method df.describe () is a function that performs a computation, and it requires parentheses to call the function.                          

Summary of the Pre-Lecture HW Session 

Initial Prompt: You asked about the concept of an initial prompt, which sets the context for a conversation or interaction.

Pandas DataFrame: You shared code to read a dataset into a Pandas DataFrame and asked how to determine the number of rows and columns. I provided the use of the .shape attribute for that purpose.

Summarizing Columns: I explained how to get summaries of the DataFrame's columns using methods like info(), describe(), and other techniques for viewing data types, unique values, and missing values.

Discrepancies Between df.shape and df.describe(): We discussed why df.describe() might analyze fewer columns and show different counts compared to df.shape, specifically focusing on the types of columns analyzed and missing values.

Attributes vs. Methods: You learned the difference between attributes (e.g., df.shape) which provide properties without parentheses, and methods (e.g., df.describe()) which perform actions and require parentheses.


Link: https://chatgpt.com/share/66e3ae32-3e88-8006-920f-f773a70d4d19

# Post- lecture

1. Definition of Summary Statistics: <br>
  a) Count: It tells you how many non-null (non-missing) values are present in each column. It counts the number of values available for statistical analysis.
  
  b) Mean: Mean is referred to as the arithmetic average and is the average value of the data points in the column. Its Calculation: mean= (sum of all the values) / count of non-missing values. df ['x'].mean().
  
  c) Standard Deviation (std): Provides a measure of the spread of data points around the mean. Low std is interpreted as the values being closer to the mean, while a higher one indicates that the values are more spread-out. df ['x'].std()
  
  d) Minimum (min): The smallest value represented in the column for a datset. df ['x'}.min ().
  
  e) 25% (1st quartile):The value below which 25% of the data fall. df['x'].quantile(0.25)
  
  f) 50% (median or 2nd quartile): The middle value in the data set--half of the data falls below this point and half above. df['x'].quantile(0.5)
  
  g) 75% (3rd quartile): The value below which 75% of the data fall. df['x'].quantile(0.75)
  
  h) Maximum (max): The largest value represented in the column for a data set. df ['x'].max().
  
  
  

7. "Across rows," you are focusing on individual observations within the data set, which are represented as rows in the DataFrame. df.dropna () can be used to remove rows that contain missing values if only a few rows are incomplete to help retain as much data as possible.

"Down columns," you are focusing on the variables within the data set. When there are many missing values in a single column, this might indicate the variable is not very informative, and using del df ['col'] to remove the entire column is an efficient strategy. 

The default df.dropna() or df.dropna (axis=0) function eliminates rows that contain any missing data.
df.dropna (axis=1) removes columns with any missing values.
df.dropna (how ="all") removes a row or column if all values are missing.
del.df ['col'] completely deletes the mentioned column in the data set.



7.1 Case: Medical dataset where each column represents important measurements for patients like blood pressure, temperature, etc. If there are a few missing measurements for a patient across many rows and columns that are highly important for the medical analysis, dropping the entire columns using del df ['col'] will result in the loss of valuable variables and information. However, if only the rows with missing data are dropped using df.dropna (), all important columns will remain available for medical analysis. 


7.2 Case: A survey data set where the majority of responses are missing for "favourite olor." Deleting the variable or column favourite colour using del def. ['col'] will not impact the statistical analysis and allow for a greater focus on the more complete columns. This is a better option than using df. dropna () to delete rows, which might provide relevant information to the analysis.   

7.3 When both del df ['col] and df.dropna () are used together, applying del df ['col'] first is important as it removes irrelevant columns and prevents the loss of important data in rows.

In [10]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
village_df = pd.read_csv(url)

# Check for missing values in each column
missing_data_before = village_df.isna().sum()
print("Missing data before cleanup:\n", missing_data_before)


Missing data before cleanup:
 row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


In [18]:
village_df

row_n  gender birthday personality             song    phrase  \
0        2    male     1-27      cranky       Steep Hill   aye aye   
1        3  female      7-2       peppy          DJ K.K.  sidekick   
2        4  female     4-21        uchi       K.K. House   snuffle   
3        6    male    10-18        lazy       Steep Hill   Ayyeeee   
4        7    male      6-9        lazy      Forest Life  it'sa me   
..     ...     ...      ...         ...              ...       ...   
386    475  female     1-31       peppy         My Place    hay-OK   
387    477    male    11-25      cranky        K.K. Song   snarrrl   
388    480  female     7-20      snooty     Soulful K.K.   tsk tsk   
389    481    male      6-7        smug         K.K. D&B     pronk   
390    483    male      3-8        lazy  Spring Blossoms     bloop   

               full_id                                                url  
0     villager-admiral  https://villagerdb.com/images/villagers/thumb/...  
1     villager-agent-s  https://villagerdb.com/images/villagers/thumb/...  
2       villager-agnes  https://villagerdb.com/images/villagers/thumb/...  
3          villager-al  https://villagerdb.com/images/villagers/thumb/...  
4     villager-alfonso  https://villagerdb.com/images/villagers/thumb/...  
..                 ...                                                ...  
386    villager-winnie  https://villagerdb.com/images/villagers/thumb/...  
387  villager-wolfgang  https://villagerdb.com/images/villagers/thumb/...  
388      villager-yuka  https://villagerdb.com/images/villagers/thumb/...  
389      villager-zell  https://villagerdb.com/images/villagers/thumb/...  
390    villager-zucker  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 8 columns]

In [33]:
village_df.dropna ()

row_n birthday personality             song    phrase            full_id  \
0        2     1-27      cranky       Steep Hill   aye aye   villager-admiral   
1        3      7-2       peppy          DJ K.K.  sidekick   villager-agent-s   
2        4     4-21        uchi       K.K. House   snuffle     villager-agnes   
3        6    10-18        lazy       Steep Hill   Ayyeeee        villager-al   
4        7      6-9        lazy      Forest Life  it'sa me   villager-alfonso   
..     ...      ...         ...              ...       ...                ...   
386    475     1-31       peppy         My Place    hay-OK    villager-winnie   
387    477    11-25      cranky        K.K. Song   snarrrl  villager-wolfgang   
388    480     7-20      snooty     Soulful K.K.   tsk tsk      villager-yuka   
389    481      6-7        smug         K.K. D&B     pronk      villager-zell   
390    483      3-8        lazy  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[380 rows x 7 columns]

7.4 I used village_df.dropna() and ended up with 380 rows and 7 columns. My rows were reduced from 391 to 380, indicating that the 11 rows that were removed had missing values, and my columns were reduced from 8 to 7, indicating that one column had missing values in all or most rows.

chat gpt summary

Here’s a summary of the session:

Dataset Loading and Inspection:

You loaded the dataset from the URL and inspected it using village_df.isna().sum() to check for missing values.
Handling Missing Data:

We discussed how to handle missing data by using df.dropna() to remove rows with missing values and del df['col'] to remove columns with excessive missing values.
Example scenarios and justifications were provided for when to use del versus df.dropna().
Removing Columns:

You encountered a KeyError when trying to remove the 'id' column using del village_df['id'] because the column might not have existed or was incorrectly specified.
Solutions were provided to verify the actual column names and correctly remove the column.
Removing Specific Rows:

Instructions were given on how to remove specific rows (e.g., rows 5 to 385) using slicing with drop() instead of dropna(), which is used for removing rows with missing values.
Effect of dropna():

After using village_df.dropna(), you observed a reduction in both the number of rows and columns. This indicated that rows with missing values were removed and potentially a column with many missing values was dropped.
Key Actions Taken:
Removed Columns: Using del or drop based on the column's presence and correct name.
Removed Rows: Used slicing to exclude specific rows.
Handled Missing Data: Applied dropna() and noted the impact on the dataset

https://chatgpt.com/share/66e638a1-03e4-8006-a94a-76cb6efa723d

8.1 The df.groupby('col1') function groups the DataFrame df using the values in ('col1'). The ["col2"].describe() function generates summary statistics for the ('col2') such as std, mean, count, 50%, 25%, etc.

In [14]:
import pandas as pd

# Load the Titanic dataset
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv')

# Group by the 'class' column and get descriptive statistics for the 'age' column
age_describe_by_class = df.groupby('class')['age'].describe()

print(age_describe_by_class)

        count       mean        std   min   25%   50%   75%   max
class                                                            
First   186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
Second  173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
Third   355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0


8.2 The count values in df.describe show the number of non-missing values for each column across the DataFrame, giving a broad overview of the data set. df.groupby("col1")["col2"].describe () counts the number of non-missing values for each subgroup, such as 'col1' and 'col2' giving a more detailed view.

8.3 While introducing the errors in my code to both ChatGPT and Google, I found working with ChatGPT easier than working with Google for searching and fixing my errors. This is because Chat GPT generated a response that was tailored to my question, explaining the error, how to fix it, and also giving me the correct code. In Google, however, I did not get a specific response for my question but only a general one about the meaning of my error. Google was also not able to correct my code but rather only tell me where the error could potentially be in my code. Furthermore, when searching in Google, I had to open different websites to get an answer to a simple question, which is less efficient than ChatGPT. 

Summary of ChatGPT Session Certainly! Here’s a summary of our session:

Error Fixes and Syntax Corrections:

You encountered various errors while working with pandas and the Titanic dataset.
Errors included NameError, KeyError, SyntaxError, and issues with URL loading.
We corrected issues with URL typos, case sensitivity in column names, and syntax errors in pandas operations.
Key Corrections:

URL for Titanic Dataset: Ensure the correct URL is used for loading the dataset. It should be https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv.
Column Names: Use the correct column names, which are case-sensitive. For example, use 'sex' instead of 'Sex'.
Syntax for groupby and Indexing: Column names should be enclosed in quotes in pandas operations. For example, use df.groupby('class')['age'].describe().
General Advice:

Always ensure that variable names and column names match the expected case.
Double-check URLs and syntax when encountering errors.
Import necessary libraries and define variables before using them.
Feel free to reach out if you have any more questions or need further assistance!

Link: 
https://chatgpt.com/share/66e7a97b-2668-8006-b3be-b93ac06648e2